### ISSUE TITLE
Integration into CPython

### ISSUE THREAD

**Issue by gvanrossum** [gvanrossum/pegen#182]
I talked to Łukasz and he's excited to see this integrated into CPython 3.9. Some of his suggestions:

- Test on the 100 most popular PyPI packages (in addition to stdlib), or possibly on all of PyPI (1TB)
- Make it opt-out (e.g. through a command line flag, environment var, and sys.get/set function)

**Comment by gvanrossum** [comment]
My own idea is that perhaps we could merge an earlier, opt-in version now (in time for the next alpha), and then starting beta1 we could change the switch to opt-out.

**Comment by lysnikolaou** [comment]
I'd really love to work on this, but it's your call since I guess that it's going to go a bit slower than if you or @pablogsal do the integration.

**Comment by pablogsal** [comment]
> I'd really love to work on this, but it's your call since I guess that it's going to go a bit slower than if you or @pablogsal do the integration.

We can certainly work together on it :)

**Comment by lysnikolaou** [comment]
> > I'd really love to work on this, but it's your call since I guess that it's going to go a bit slower than if you or @pablogsal do the integration.
> 
> We can certainly work together on it :)

That'd be great!

**Comment by pablogsal** [comment]
I suppose that we need to finish first #180.

**Comment by pablogsal** [comment]
@gvanrossum Do you think moving definitely the parser to CPython will require a PEP?

**Comment by gvanrossum** [comment]
I'm not sure. It certainly needs multiple people to think about the ramifications. The accelerated sunset of the parser module is one concern (though covered by PEP 594, which specifically says it will be removed in 3.9 to be replaced with a more advanced parser).

**Comment by lysnikolaou** [comment]
> * Test on the 100 most popular PyPI packages (in addition to stdlib), or possibly on all of PyPI (1TB)

I would like to start setting this up. Where do we get what the 100 most popular PyPI packages are? Is https://hugovk.github.io/top-pypi-packages/ ok?

**Comment by gvanrossum** [comment]
Heh, that sounds like a good start!

**Comment by lysnikolaou** [comment]
Any idea on where to run these tests?

Downloading the source of the top 100 PyPI packages should not exceed 100MB of memory, although it could. Still, uploading all the compressed files on GitHub doesn't seem like a good idea. Even more so, if we want to scale this to testing all of PyPI.

**Comment by gvanrossum** [comment]
Write a script that downloads them all, another script that runs the tests,
and run it manually?

Also, have you thought about profiling?
-- 
--Guido (mobile)

**Comment by lysnikolaou** [comment]
I tried out some things but quickly gave up, since my experience with profiling is very very limited. Also, the fact that we try to profile a C Extension for Python made it difficult to find good online resources.

I tried out using [yep](https://pypi.org/project/yep/), which didn't prove useful at all. I also tried out airspeed velocity, which I gave up pretty quickly as well, since it wasn't that easy to set up. I can have another look to that, if you think we should invest more time in it.

**Comment by gvanrossum** [comment]
Without profiling we’re optimizing in the dark — we gotta do it.
-- 
--Guido (mobile)

**Comment by lysnikolaou** [comment]
> Without profiling we’re optimizing in the dark — we gotta do it. -- --Guido (mobile)

Surely! I was talking specifically about airspeed velocity.

**Comment by ethanhs** [comment]
I've found cProfile with https://pypi.org/project/pyprof2calltree/ to be rather nice for pure Python code. 

For C extensions I've had luck with https://github.com/benfred/py-spy, but ymmv.

I'm happy to help with testing this btw, I've been thinking fuzzing the parser would be a good idea too.

I recall reading Donald Stufft has previously asked people to request a snapshot of PyPi rather than try to download the entire thing themselves. (I dm'd him on Twitter asking about this).

**Comment by gvanrossum** [comment]
The code to profile is all fiddly C code, so maybe we should give py-spy a spin.

Fuzzing the parser is also important but I think of less priority: I'm pretty confident that we reject most invalid programs. Pointing the caret at the same position is a different story though (sometimes the PEG parser may read ahead more before it backtracks, although when I tried it, it was pretty accurate).

I don't have room on my hard drive for the full snapshot (1TB???), and honestly I think we'll do fine with the top 100 or so (maybe 100 MB?)

**Comment by lysnikolaou** [comment]
I didn't have much luck with py-spy. On their homepage a `--native` command line option is mentioned, in order to profile C extensions, but when I actually installed the package and ran it there was no such option. I also tried running py-spy without the `--native` flag, but then the prof file did not contain our C functions, as expected.

**Comment by ethanhs** [comment]
@lysnikolaou --native is an argument of the record subcommand. I went ahead and profiled both xxl and the pypi suite (since what better profile than real code!). Here are the results:

- xxl:
   - raw: https://0xe.me/static/profile-xxl
   - flamegraph: https://0xe.me/static/profile-xxl.svg

- PyPi
  - raw data: https://0xe.me/static/profile-pypi
  - flamegraph: https://0xe.me/static/profile-pypi.svg

Note you can interact with these svgs and select elements to zoom in on them. You may want to read www.brendangregg.com/flamegraphs.html if you are unfamiliar with flamegraphs.

**Comment by lysnikolaou** [comment]
Oh, thanks a lot for doing this! This is really useful.

Couple of questions:
1) How many PyPI packages did you profile pegen against?
2) Could you maybe explain how you got these results? What command/script did you use?

**Comment by gvanrossum** [comment]
Also, flamegraphs require some explanation if you've not seen them before
-- they're not what they seem!
http://www.brendangregg.com/flamegraphs.html

**Comment by ethanhs** [comment]
Certainly!

For xxl I ran:

```
py-spy record --native -o profxxl.svg -- python -c "from pegen import parse; parse.parse_file('data/xxl.txt', mode=2)"
```

for pypi, I just ran the default download/test script, so 100. The command I used (after downloading) was:

```
py-spy record --native -o profile.svg -- python scripts/test_pypi_packages.py
```

You can add `--format raw` before the `--` in each command to get the raw output.

Also I have a lot of storage on my server so I will probably run the pypi test script with a lot more packages (I haven't heard back from Donald Stufft).

**Comment by ethanhs** [comment]
Btw, I multiprocessed the `test_pypi_packages.py` script and am running it on all 4000 packages listed. It seems to be coming up with quite a few differences in the AST. I'll post the full log once it is done (later tonight maybe?)

**Comment by gvanrossum** [comment]
Note that once a difference is found, the utility runs the rendered AST through Black, which may be quadratic (at least it takes forever for medium-sized files).

Also, you'll get lots of diffs for decorators with at least three dotted names (`@a.b.c` or more) -- we are fixing this upstream. If you're not testing with this patch, start over! https://bugs.python.org/issue39579

**Comment by ethanhs** [comment]
Ah, yeah I realized that. Also realized I should probably pull in the f strings branch for pegen. 

So far I have found a definite bug: decorated async functions aren't parsed into AsyncFunctionDef nodes.

e.g.
```
@app
async def f():...
```

I will probably start over tomorrow.

**Comment by lysnikolaou** [comment]
Good catch, @ethanhs! Opened #194 for this.

**Comment by ethanhs** [comment]
Okay I've started things up with @pablogsal's fstring branch and a patched CPython. I will start opening issues as I notice them (and post the full log here once it completes, I won't get to every issue). It seems there are issues arising in around 1/20 packages (a rough estimate), so its good we are doing this :)

**Comment by gvanrossum** [comment]
Thank you so much for doing this! This is immensely valuable. So far we've
just tested with the stdlib, which clearly doesn't exercise *every*
syntactic construct. (Also thanks @apalala for reminding us of the
importance of doing this!)

**Comment by ethanhs** [comment]
No worries! The total size of packages listed in the JSON is under 10GB so it should be easy to reproduce (depending on how long it runs perhaps 😛  )

**Comment by ethanhs** [comment]
Okay so it took about an hour to get through something like 3800 packages, but then it really slowed down once it hit `kubernetes` and ran black on it, so it took a bit longer than that.

Here is the full (125673 line, 6MB!) log, mostly errors for the issues I opened already, but feel free to look through if you want, I probably missed some issue.


[pypitest.log](https://github.com/gvanrossum/pegen/files/4177516/pypitest.log)

**Comment by pablogsal** [comment]
@ethanhs This does not include the fixes in #199 and #198 no?

**Comment by ethanhs** [comment]
@pablogsal no it doesn't have those fixes, I found those by running this test :)

I can probably skip kubernetes and run it again if people want a smaller log to deal with.

**Comment by lysnikolaou** [comment]
@ethanhs Thanks a lot! Really great work!

Could you maybe run the tests once more with #198 and #199 (maybe without Kubernetes)?

**Comment by ethanhs** [comment]
Sure thing! Hopefully that cuts down a lot of the noise!

**Comment by ethanhs** [comment]
Sorry I've been busy (starting a new job) but I'll try pulling changes and testing things tomorrow.

**Comment by ethanhs** [comment]
Okay I got a chance to do this now! I found #209. The only other issue I noticed looks to be #205.

Here is the full log, note viewing it raw or downloaded is probably best to spot the differences in non-ASCII characters:

https://gist.github.com/ethanhs/02734c6c1e58b3fa58646259f6915462

It appears that about 3800 packages had their ASTs diff'd. The average parse speed was 115875 lines/sec.

**Comment by lysnikolaou** [comment]
Could someone please sum up the findings one can extract from the flame graphs? I've been trying to study them, but, even after reading the above mentioned article, I can't really understand what the most important results are.

**Comment by lysnikolaou** [comment]
@gvanrossum @pablogsal Do you think that it's maybe time to start working on the integration? I don't really have a clue how much work this is, but do you think we could be there by alpha3, which is scheduled for 25.03?

**Comment by pablogsal** [comment]
> Do you think that it's maybe time to start working on the integration? I don't really have a clue how much work this is, but do you think we could be there by alpha3, which is scheduled for 25.03?

I was under the assumption that we needed to present this to the other core devs in the language summit before proceeding to integrate this into the core as I assume some of them may find the integration surprising otherwise. I may be wrong or missing something, though.

**Comment by gvanrossum** [comment]
I think that what you're missing is that if it is met with approval (or at worst "meh") at the summit, it will be a big project to do the integration. If we started a "WIP" PR with the integration work now we would be able to do things quicker once we have the green light.

I don't know if it was written up already, but I think we decided that there should be a command-line flag to Python to enable this. In the last alpha and in the first beta the default should be the old parser. Then in later betas the default should be the new parser. This way we may be able to keep the old 'parser' module around in 3.9 (since a little birdie told me that PEP 594, which proposes to kill 'parser' in 3.9, may be deferred to 3.10).

I'd also like to have an explicit way to invoke the new parser even when the command-line flag specifies the old parser (e.g. for unit-testing the new parser). This should just be importing something (or perhaps an alternative API in the ast module).

 Also, hat happened to performance?!! It's almost twice as fast now. If that's true, that's fantastic!

**Comment by pablogsal** [comment]
>  If we started a "WIP" PR with the integration work now we would be able to do things quicker once we have the green light.

Oh, I see. I understand now and I agree that we should indeed start as soon as possible. I see three ways we could work in the integration:

1. Open a WIP/draft PR. This requires less setup but we would need to push to the same branch (the fork of who opens the PR) in the first place. The problem that I see with this one is that reviewing this PR (thinking about us doing to review) will become more and more difficult as the PR grows bigger.

2. Open a branch in CPython (new-parser or something similar), do PRs against there to develop the integration and merge that branch into master once we want to incorporate pegen finally. This has the advantage that we can do smaller PRs.

3. Fork CPython somewhere and all of us push to that fork to develop the integration and when we finish, we do the final PR against CPython from that fork. This has the advantage that we can push directly to the fork if we need to/we prefer doing that directly.

What do you prefer? Do you see another workflow you prefer?

> I don't know if it was written up already, but I think we decided that there should be a command-line flag to Python to enable this. In the last alpha and in the first beta the default should be the old parser. Then in later betas the default should be the new parser.

:+1: 

> I'd also like to have an explicit way to invoke the new parser even when the command-line flag specifies the old parser (e.g. for unit-testing the new parser). This should just be importing something (or perhaps an alternative API in the ast module).

I think a new temporal/private API in the AST module would be the best as we probably would like to get AST objects to do the unit-testing for this case.

> Also, what happened to performance?!! It's almost twice as fast now. If that's true, that's fantastic!

In the last 20 commits, we have evolved from `5.5333` seconds to `3.091` seconds in `simpy_cpython` in my system. :tada:

**Comment by gvanrossum** [comment]
I like having a branch we can all push to. That's (nearly) what I've been using for the PEP 585 work (the branch is owned by my own cpython fork and I just accept PRs against the branch in my fork).

The speed is now much closer to that of the old parser (within 20% probably). Did this come from [profiling](https://github.com/gvanrossum/pegen/issues/181) or from intuition alone?

The next challenge is memory usage. (If you want to discuss specific ideas about that, one of you should open a new issue for that topic.)

**Comment by pablogsal** [comment]
Opened https://github.com/gvanrossum/pegen/issues/220 to track the memory usage

**Comment by lysnikolaou** [comment]
> What do you prefer? Do you see another workflow you prefer?

You both know best, what the best workflow is. 

> The speed is now much closer to that of the old parser (within 20% probably). Did this come from profiling or from intuition alone?

Whatever changes I did, only came from intuition. As stated above, I unfortunately could not read out results from the published flamegraphs. 

> In the last 20 commits, we have evolved from 5.5333 seconds to 3.091 seconds in simpy_cpython in my system. 🎉

Most of the speedup must be related to #217, which just disables AST creation, in case it is not needed.

**Comment by gvanrossum** [comment]
Regarding the flame graphs:

0. One nice thing is that this shows both Python and C stack frames!
1. It's upside down from the example in the [article](http://www.brendangregg.com/flamegraphs.html).
2. Ignore the colors, they are just chosen to differentiate frames.
3. The X axis is NOT the flow of time! It is how often a given frame was seen. So `main` (like `<all>`, the root) extends over the full width.
4. The Y axis represents the stack (growing down). So e.g. `<all>` calls `main` calls `run_tests` calls `parse_directory` and so on.
5. You can click on an item and get a "zoomed" version of the graph. (Scroll to top for a "Reset zoom" button -- but clicking "all" works too.)

One issue is that the line numbers quickly lose their meaning as the code generator changes, so in order to accurately understand the graphs that Ethan posted you'd need to check out the repo from around the date that Ethan ran it.

**Comment by gvanrossum** [comment]
> Most of the speedup must be related to #217, which just disables AST creation, in case it is not needed.

I actually doubt that. #217 just changes the testing methodology to be more accurate. But I used a different methodology (running `make time`) which doesn't even use the script you changed in #217, and it still is much faster. So give your intuition some credit! :-)

**Comment by pablogsal** [comment]
I have created an org and forked cpython from there and I have invited you as admins of the org:

https://github.com/we-like-parsers/cpython

This way all three of us have control over branches, configuration, CI....etc We can set up a branch (or multiple ones) in that fork to work on the integration. I think that this allows us to have more control than using a personal fork.

**Comment by lysnikolaou** [comment]
> Please, tell me if you are comfortable with this or you prefer a different setup :)

I definitely like this approach. Thanks for setting everything up!

**Comment by gvanrossum** [comment]
Cool! I will probably use a separate local repo tied to this, with python/cpython as "upstream" remote.

I've enabled issues so we can start creating tasks.

I'm not sure whether we should work in the master branch or create a separate "pegen" branch -- suggestions?

**Comment by pablogsal** [comment]
> I'm not sure whether we should work in the master branch or create a separate "pegen" branch -- suggestions?

If we let master mirror the CPython master, we can explicitly merge that branch into our "pegen" branch if we ever need to synchronize with the latest CPython. This is better than rebasing as this allows us to not need to force push to do the updates, forcing everyone to reset locally.

On the other hand, if we don't plan to rebase into the latest CPython until the end, we can just use 'master' directly to simplify the workflow and push there or create PRs against there if we want reviews.

**Comment by lysnikolaou** [comment]
I was just about to say what @pablogsal said, but it's totally up to you both, since I don't really know if we are going to have to synchronise with CPython master.

**Comment by lysnikolaou** [comment]
> I've enabled issues so we can start creating tasks.

Also, we could create a project board for this. I've never used them on Github, but this seems like a nice opportunity to do so.

**Comment by gvanrossum** [comment]
OK, let's use a branch named 'pegen'. I will create and push it.

**Comment by gvanrossum** [comment]
Note, I have changed the default branch to 'pegen'. Hopefully this works. Also, everybody should start watching this repo.

**Comment by lysnikolaou** [comment]
This can probably be closed, since the initial integration is done.
